proposals/003-metric-improvenents.md

```diff
@@ -0,0 +1,262 @@
+
+### Specific short-comings/weaknesses
+
+* *1 - these metrics record only the **requests**, there's nothing analogous that records the responses.
+* *1 - these metrics aren't discriminated by ApiKey (etc.). This limits their usefulness for problem-solving.
+* *2 - the metric name is quite misleading - nothing tells the user that it is only recording the sizes of RPCs that are actually **decoded**.
+* *2 - there are no quantised buckets configured for this metric. it gives the user count/sum/max values only. it would be
+  hard to use this metric to understand something about produce requests or fetch responses.
+* *1 / *2 - we have contradictory meanings for the `flowing` label.  in *2, requests flow `UPSTREAM`, responses flow `DOWNSTREAM`.
+  In *1 requests flow `DOWNSTREAM` which makes no sense.
+* *3 - downstream errors does not record account for errors such as TLS negotiation errors or failures to determine virtual cluster or broker.  
+
+### General short-comings/weaknesses
+
+* The metrics endpoint populates `HELP` for each metric, but this is populated with nothing more than the metric's name.
+  We ought to be giving a clear description of the metric.
+* Some of the label names don't follow Prometheus naming conventions.  Prometheus recommends snake case whereas we have used camel case in some cases.
+
+## Use Cases
+
+In general, metrics exist to help a user understand the health and performance of the proxy. They help inform
+capacity planning, drive system alerts reporting abnormal conditions and help with problem-solving (things like
+diagnosing failures or system slow-downs).
+
+Introducing a proxy to a Kafka system will increase latency of the system. Users will need a mechanism to understand how
+much latency is being added for their particular use-cases and have the ability to monitor this latency over time.
+
+Kroxylicious Filters themselves have the potential to alter behaviour of the Kafka system.  Users will need insights into
+performance and other issues that might result from filter behaviour. Specifically:
+
+* filters have the ability to change the size of the RPCs as they traverse the proxy (record encryption)
+* filters have the ability to change the flow of messages within the Kafka system. Specifically, they can short-circuit
+  requests by producing their own responses (the broker never sees the request) and introduce their own requests.
+* filters can perform asynchronous work.  When asynchronous work occurs, the proxy stops reading the downstream or upstream.
+
+## Proposal
+
+## Summary
+
+The following metrics will be added.  They will each be described in more detail later in this section.
+
+| Metric                                           | Type         | Labels                                                                                | Description                                                                             |
```

Comment:
Do we really need to prefix every metric with `kroxylicious`? It's long, difficult to type and difficult (for some) to pronounce.